### DISCUSSION TITLE
Support for other versioning schemes

### DISCUSSION THREAD

**Discussion by NetOpWibby** [npm/rfcs#348] [RRFC]
Hi y'all.

I favor [ChronVer](https://chronver.org) over SemVer and it's irritating to have my flavor of versioning mangled into SemVer. At the very least, I would appreciate a flag that'd let me bypass SemVer checks for certain packages (i.e., my own).

This would be a user-configurable thing, with SemVer as default to keep 99% of the userbase satisfied.

**Comment by ljharb** [comment]
That seems like a ton of complexity to add. How many packages use this versioning scheme? How popular are these packages?

The entire npm ecosystem functions on semver; it seems like if you want a different versioning scheme, a different ecosystem would be more appropriate.

**Reply by NetOpWibby** [reply]
Which other ecosystems exist? As far as I know, npm is THE ecosystem.

I'm not sure what is involved here but if push comes to shove I suppose I could figure out how to create a bridge of some sort but I'd rather not.

All I'm suggesting is choice.

**Reply by ljharb** [reply]
Sure, and semver is THE versioning system for it. Using an alternate versioning system is the same level of impact as using an alternate ecosystem.

**Reply by NetOpWibby** [reply]
So npm cannot ever support another versioning system? Ever?

**Reply by ljharb** [reply]
I don’t see any way to do so that wouldn’t cause compatibility issues with every extant package - but more importantly, i don’t see any value to be had by doing so.

**Reply by NetOpWibby** [reply]
I am not suggesting replacing SemVer. At all. Why would compatibility issues occur?

**Reply by dominykas** [reply]
@ljharb I don't think I agree with your position the people should go play elsewhere.

I also don't think it's true, by far, that semver is the only versioning schema that's used by the ecosystem published in npm. TypeScript explicitly does _not_ follow semver.

I think there is value to be had in allowing packages to specify explicitly that they are not following semver. They should, but they don't have to.

I also think there is value to be had in broadening and accepting more people and packages into the ecosystem. Of course, in this specific case, I'm sure there's implications, because there's lots of existing assumptions about what are the allowed version specifiers. I'm also sure that it's possible to solve this. And I'm most certain that it's unfair and unjust to tell people to go look for another ecosystem, especially because they just might do that, and that would be a loss.

At the very least, the ideas are worth discussing.

**Reply by ljharb** [reply]
There’s two aspects here: following semver, which certainly not everyone does, and using semver-compatible versions, which is the part that i don’t think can be changed.

**Comment by dominykas** [comment]
> let me bypass SemVer checks for certain packages

Which checks do you mean by this specifically?

**Reply by NetOpWibby** [reply]
What prompted me opening this discussion was an issue that arose with npm telling me a version number was invalid. I'm not quite sure how I managed to publish a weird version to my Verdaccio instance a few months ago but a fresh `npm i` and explicitly choosing a version to install got me around it.

**Reply by dominykas** [reply]
What was the invalid version?

**Reply by NetOpWibby** [reply]
`2021.2.1-1.1`. The one before it was `2021.2.11`, which is what SemVer converts it to from `2021.02.11`.

**Reply by dominykas** [reply]
Ah, right, the leading zero.

I personally don't see a major problem if leading zeroes were allowed by npm in some form (they are explicitly forbidden by the semver standard, though).
